app/src/main/java/ee/ioc/phon/android/speechutils/ContinuousRawAudioRecorder.java

```diff
@@ -0,0 +1,215 @@
+        SessionStartPointer.setRecordingBufferLengthMillis(recordingBufferLengthMillis);
+
+        try {
+            int bufferSize = getBufferSize();
+            int framePeriod = bufferSize / (2 * RESOLUTION_IN_BYTES * CHANNELS);
+            createRecorder(audioSource, sampleRate, bufferSize);
+            createBuffer(framePeriod);
+            setState(State.READY);
+        } catch (Exception e) {
+            if (e.getMessage() == null) {
+                handleError("Unknown error occurred while initializing recorder");
+            } else {
+                handleError(e.getMessage());
+            }
+        }
+    }
+
+    public ContinuousRawAudioRecorder(int sampleRate) {
+        this(DEFAULT_AUDIO_SOURCE, sampleRate, DEFAULT_BUFFER_LENGTH_IN_MILLIS);
+    }
+
+    public ContinuousRawAudioRecorder() {
+        this(DEFAULT_AUDIO_SOURCE, DEFAULT_SAMPLE_RATE, DEFAULT_BUFFER_LENGTH_IN_MILLIS);
+    }
+
+    public String getWsArgs() {
+        return "?content-type=audio/x-raw,+layout=(string)interleaved,+rate=(int)" + getSampleRate() + ",+format=(string)S16LE,+channels=(int)1";
+    }
+
+    public ContinuousRawAudioRecorder setSessionStartPointer(SessionStartPointer sessionStartPointer) {
+        mSessionStartPointer = sessionStartPointer;
+        return this;
+    }
+
+    private int calculateNumOfSamplesToGoBack(int startPos) {
+        // if the consumed session is not the same as the recorded session
+        // get the data from the beginning of the buffer/desired session start pointer
+        if (!isRecordedSessionSameAsConsumedSession()) {
+            Log.i(LOG_FILTER + "Recorded session and consumed session are NOT the same. Grabbing the data from the session start position");
+            return getNumOfSamplesIn(mSessionStartPointer.getSessionStartPointerMillis());
+        }
+
+        Log.i(LOG_FILTER + "Recorded session and consumed session are the same. Working according to the consumed pointer");
+        if (startPos == getLength()) {
+            Log.i(LOG_FILTER + "Consumed session pointer still points to the recorded session pointer. Nothing to do or to return");
+            return -1;
+        }
+
+        int numOfSamplesToGoBack = getLength() - startPos;
+        if (numOfSamplesToGoBack < 0) // cyclic buffer case
+            numOfSamplesToGoBack += mRecording.length;
+
+        Log.i(LOG_FILTER + "Consumed session pointer is behind the recorded session pointer by: " + numOfSamplesToGoBack + " samples");
+        return numOfSamplesToGoBack;
+    }
+
+    @Override
+    protected byte[] getCurrentRecording(int startPos) {
+
+        int numOfSamplesToGoBack = calculateNumOfSamplesToGoBack(startPos);
+        if (numOfSamplesToGoBack <= 0) {
+            Log.i(LOG_FILTER + "There are no samples that we need to take from the recording");
+            return null;
+        }
+
+        byte[] buffer = new byte[numOfSamplesToGoBack];
+        int potentialStartSample = getLength() - numOfSamplesToGoBack;
```

Comment:
this is a potential start sample for consumption since if the recording pointer is lower than the consuming pointer, this value is negative - which means that the buffer copy should happen from two different places in the recording buffer. See L.146 for the data copy split